### ISSUE TITLE
Let AbilityBuilder DSL accept Constructor as subject

### ISSUE THREAD

**Issue by icedtoast** [stalniy/casl#58]
It would be handy if the AbilityBuilder supported the constructor function in lieu of the subject name when creating rules with `can`. After all, it already has the `subjectName` option passed in.

For example:
```
import { AbilityBuilder } from '@casl/ability'
import Post from './post'

function defineAbilitiesFor(user) {
  return AbilityBuilder.define((can, cannot) => {
    if (user.isAdmin()) {
      can('manage', Post)
    } else {
      can('read', Post)
    }
  })
}
```

I am happy to submit a pull request with this feature.

**Comment by stalniy** [comment]
Hi

I’m ok with that. But you need to make sure that `subject` field in resulting rules produced by DSL is a string. 

Because `Ability` creates an index (plain object) for that rules in order to find them quickly. I still don’t want to use `Map` Because this will require to add polyfill on front end side

Waiting for PR :)

**Comment by stalniy** [comment]
Don’t forget to update typescript definitions

**Comment by stalniy** [comment]
available in `@casl/ability@2.3.0`! Thanks for contribution!
